src/internal/kopia/wrapper.go

```diff
@@ -710,3 +713,184 @@ func (w Wrapper) setMaintenanceParams(
 
 	return nil
 }
+
+// SetRetentionParameters sets configuration values related to immutable backups
+// and retention policies on the storage bucket. The minimum retention period
+// must be >= 24hrs due to kopia default expectations about full maintenance.
+func (w *Wrapper) SetRetentionParameters(
+	ctx context.Context,
+	retention repository.Retention,
+) error {
+	if retention.Mode == nil && retention.Duration == nil && retention.Extend == nil {
+		return nil
+	}
+
+	// Somewhat confusing case, when we have no retention but a non-zero duration
+	// it acts like we passed in only the duration and returns an error about
+	// having to set both. Return a clearer error here instead.
+	if ptr.Val(retention.Mode) == repository.NoRetention && ptr.Val(retention.Duration) != 0 {
+		return clues.New("retention duration must be 0 or unset if disabling retention")
+	}
+
+	dr, ok := w.c.Repository.(repo.DirectRepository)
+	if !ok {
+		return clues.New("unable to get valid handle to repo").WithClues(ctx)
+	}
+
+	blobCfg, params, err := getRetentionConfigs(ctx, dr)
+	if err != nil {
+		return clues.Stack(err)
+	}
+
+	// Update blob config information.
+	blobChanged, err := w.setRetentionMode(retention.Mode, blobCfg)
+	if err != nil {
+		return clues.Wrap(err, "setting retention mode")
+	}
+
+	if retention.Duration != nil && blobCfg.RetentionPeriod != *retention.Duration {
+		blobCfg.RetentionPeriod = *retention.Duration
+		blobChanged = true
+	}
+
+	// Update maintenance config information.
+	var maintenanceChanged bool
+
+	if retention.Extend != nil && params.ExtendObjectLocks != *retention.Extend {
+		params.ExtendObjectLocks = *retention.Extend
+		maintenanceChanged = true
+	}
+
+	// Check the new config is valid.
+	if blobCfg.IsRetentionEnabled() {
+		if err := maintenance.CheckExtendRetention(ctx, *blobCfg, params); err != nil {
+			return clues.Wrap(err, "invalid retention config")
+		}
+	}
+
+	return clues.Stack(persistRetentionConfigs(
+		ctx,
+		dr,
+		blobCfg,
+		blobChanged,
+		params,
+		maintenanceChanged,
+	)).OrNil()
+}
+
+func getRetentionConfigs(
+	ctx context.Context,
+	dr repo.DirectRepository,
+) (*format.BlobStorageConfiguration, *maintenance.Params, error) {
+	blobCfg, err := dr.FormatManager().BlobCfgBlob()
+	if err != nil {
+		return nil, nil, clues.Wrap(err, "getting storage config").WithClues(ctx)
+	}
+
+	params, err := maintenance.GetParams(ctx, dr)
+	if err != nil {
+		return nil, nil, clues.Wrap(err, "getting maintenance config").WithClues(ctx)
+	}
+
+	return &blobCfg, params, nil
+}
+
+func persistRetentionConfigs(
+	ctx context.Context,
+	dr repo.DirectRepository,
+	blobCfg *format.BlobStorageConfiguration,
+	blobChanged bool,
+	params *maintenance.Params,
+	maintenanceChanged bool,
+) error {
+	// Persist changes.
+	if !blobChanged && !maintenanceChanged {
+		return nil
+	}
+
+	mp, err := dr.FormatManager().GetMutableParameters()
+	if err != nil {
+		return clues.Wrap(err, "getting mutable parameters")
```

Comment:
This is needed for many of the errors in this func.
```suggestion
		return clues.Wrap(err, "getting mutable parameters").WithClues(ctx)
```